consensus/polybft/consensus_runtime.go

```diff
@@ -0,0 +1,851 @@
+			return nil, err
+		}
+
+		// blockHeader, ok := c.config.blockchain.GetHeaderByNumber(blockHeader.Number - 1)
+		_, _ = c.config.blockchain.GetHeaderByNumber(blockHeader.Number - 1)
+	}
+
+	// since we need to calculate uptime for the last block of the previous epoch,
+	// we need to get the validators for the that epoch from the smart contract
+	// this is something that should probably be optimized
+	if lastBlockInPreviousEpoch > 0 { // do not calculate anything for genesis block
+		for i := 0; i < uptimeLookbackSize; i++ {
+			validators, err := c.config.polybftBackend.GetValidators(blockHeader.Number-2, nil)
+			if err != nil {
+				return nil, err
+			}
+
+			if err := calculateUptimeForBlock(blockHeader, validators); err != nil {
+				return nil, err
+			}
+
+			// blockHeader, ok := c.config.blockchain.GetHeaderByNumber(blockHeader.Number - 1)
+			_, _ = c.config.blockchain.GetHeaderByNumber(blockHeader.Number - 1)
+		}
+	}
+
+	return uptimeCounter, nil
+}
+
+// setIsActiveValidator updates the activeValidatorFlag field
+func (c *consensusRuntime) setIsActiveValidator(isActiveValidator bool) {
+	if isActiveValidator {
+		atomic.StoreUint32(&c.activeValidatorFlag, 1)
+	} else {
+		atomic.StoreUint32(&c.activeValidatorFlag, 0)
+	}
+}
+
+// isActiveValidator indicates if node is in validator set or not
+func (c *consensusRuntime) isActiveValidator() bool {
+	return atomic.LoadUint32(&c.activeValidatorFlag) == 1
+}
+
+// getPendingBlockNumber returns block number currently being built (last built block number + 1)
+func (c *consensusRuntime) getPendingBlockNumber() uint64 {
+	return c.lastBuiltBlock.Number + 1
+}
+
+// isEndOfEpoch checks if an end of an epoch is reached with the current block
+func (c *consensusRuntime) isEndOfEpoch(blockNumber uint64) bool {
+	return isEndOfPeriod(blockNumber, c.config.PolyBFTConfig.EpochSize)
+}
+
+// isEndOfSprint checks if an end of an sprint is reached with the current block
+func (c *consensusRuntime) isEndOfSprint(blockNumber uint64) bool {
+	return isEndOfPeriod(blockNumber, c.config.PolyBFTConfig.SprintSize)
+}
+
+// getSystemState builds SystemState instance for the most current block header
+func (c *consensusRuntime) getSystemState(header *types.Header) (SystemState, error) {
+	provider, err := c.config.blockchain.GetStateProviderForBlock(header)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.config.blockchain.GetSystemState(c.config.PolyBFTConfig, provider), nil
+}
+
+// isEndOfPeriod checks if an end of a period (either it be sprint or epoch)
+// is reached with the current block (the parent block of the current fsm iteration)
+func isEndOfPeriod(blockNumber, periodSize uint64) bool {
+	return blockNumber%periodSize == 0
+}
+
+// getQuorumSize returns result of division of given number by two,
+// but rounded to next integer value (similar to math.Ceil function).
+func getQuorumSize(validatorsCount int) int {
+	return (validatorsCount + 1) / 2
+}
```

Comment:
What does this method (that divides by 2) have to do with Quorum?

Reply:
This was just some rather temporal, dummy value. [Voting power PR](https://github.com/0xPolygon/pbft-consensus/pull/71/files#diff-b208c241c10c010125c6a1c4f43ac407c5fb7b363bab10a050e5681019875478R791) on pbft consensus, introduced helper function which calculates quorum size and which was intended to be used by v3 client. However PR which includes such change on the v3 client, never got chance to be merged :cry: (https://github.com/maticnetwork/v3/pull/223).

Hopefully go-ibft implementation provides some `QuorumSize` helper function which can be utilized on the client :slightly_smiling_face: